File: java/pe/src/ktn/pe/Pe7.java
package ktn.pe;

import java.util.Arrays;

public class Pe7 implements Pe {
    private int index = 10001;

    public static int pe7(int n) {
        if (n < 1) {
            return 0;
        }

        double times = n < 50 ? 5.0 : Math.log(n) + 2.0;
        int limit = (int) (n * times);

        boolean[] s = new boolean[limit + 1];
        Arrays.fill(s, true);
        s[0] = false;
        s[1] = false;

        int sq = (int) Math.sqrt(limit);
        for (int i = 2; i < sq + 1; ++i) {
            for (int j = i * i; j < limit + 1; j += i) {
                s[j] = false;
            }
        }

        int prime = 0;
        for (int i = 0;;) {
            if (s[prime]) {
                ++i;
                if (i >= n) {
                    break;
                }
            }
            ++prime;
        }

        return prime;
    }

    @Override
    public void setArgs(String[] args) {
        if (args != null && args.length > 0) {
            try {
                index = Integer.parseInt(args[0]);
            } catch (NumberFormatException e) {
                //
            }
        }
    }

    @Override
    public void solve() {
        System.out.println(pe7(index));
    }

}

File: kt/pe/src/main/kotlin/ktn/pe/Pe7.kt
package ktn.pe

import java.util.Arrays

class Pe7 : Pe {
    private var index = 10001

    override val problemNumber: Int
        get() = 7

    fun pe7(n: Int): Int {
        if (n < 1) {
            return 0
        }

        val times = if (n < 50) { 5.0 }
                    else { Math.log(n.toDouble()) + 2.0 }
        val limit = (n * times).toInt()

        val s = BooleanArray(limit + 1)
        Arrays.fill(s, true)
        s[0] = false
        s[1] = false

        val sq = Math.sqrt(limit.toDouble()).toInt()
        for (i in 2 until sq + 1) {
            var j = i * i
            while (j < limit + 1) {
                s[j] = false
                j += i
            }
        }

        var prime = 0
        var i = 0
        while (true) {
            if (s[prime]) {
                ++i
                if (i >= n) {
                    break
                }
            }
            ++prime
        }

        return prime
    }

    override fun setArgs(args: Array<String>) {
        if (args.isNotEmpty()) {
            index = args[0].toInt()
        }
    }

    override fun solve() {
        System.out.println(PeUtils.format(this, pe7(index)))
    }

    override fun run() {
        solve()
    }

}
